File: app/src/main/java/com/amperas17/mangaedenapp/ui/mangalist/viewmodel/MangaListViewModel.java
package com.amperas17.mangaedenapp.ui.mangalist.viewmodel;

import android.arch.lifecycle.LiveData;

import com.amperas17.mangaedenapp.data.MangaListRepository;
import com.amperas17.mangaedenapp.model.manga.Manga;
import com.amperas17.mangaedenapp.model.Resource;
import com.amperas17.mangaedenapp.ui.BaseViewModel;


import java.util.ArrayList;

public class MangaListViewModel extends BaseViewModel<ArrayList<Manga>> implements FindHandler.Caller {

    private MangaListRepository repository;
    private FindHandler handler;

    private ArrayList<Manga> mangaListAll = new ArrayList<>();

    public LiveData<Resource<ArrayList<Manga>>> getResource() {
        if (repository == null) {
            repository = new MangaListRepository(this);
        }
        if (handler == null) {
            handler = new FindHandler(this);
        }
        return super.getResource();
    }

    private void loadData() {
        repository.callData();
    }

    public void startLoading() {
        super.startLoading();
        loadData();
    }

    public void stopLoading() {
        super.stopLoading();
        repository.cancelDataRequest();
    }

    @Override
    public void onGetData(ArrayList<Manga> mangaList) {
        updateMangaListAll(mangaList);
        super.onGetData(mangaList);
    }

    private void updateMangaListAll(ArrayList<Manga> mangaList) {
        mangaListAll.clear();
        mangaListAll.addAll(mangaList);
    }


    public void findData(final String pattern) {
        if (!mangaListAll.isEmpty()) {
            if (pattern.isEmpty()) {
                postSearchResult(mangaListAll);
            } else {
                runSearchInNewThread(pattern);
                isLoading = true;
            }
        }
    }

    @Override
    public void postSearchResult(ArrayList<Manga> mangaList) {
        super.onGetData(mangaList);
    }

    private void runSearchInNewThread(final String pattern) {

        Runnable runnable = new Runnable() {
            @Override
            public void run() {
                handler.setResultList(MangaListUtils.findForMatchesInTitle(pattern, mangaListAll));
                handler.sendEmptyMessage(0);
            }
        };

        Thread thread = new Thread(runnable);
        thread.start();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        handler.removeMessages(0);
    }
}

File: app/src/main/java/com/amperas17/mangaedenapp/ui/mangalist/viewmodel/MangaListViewModel.kt
package com.amperas17.mangaedenapp.ui.mangalist.viewmodel

import com.amperas17.mangaedenapp.data.MangaListRepository
import com.amperas17.mangaedenapp.model.manga.Manga
import com.amperas17.mangaedenapp.ui.BaseViewModel


import java.util.ArrayList

class MangaListViewModel : BaseViewModel<ArrayList<Manga>>(), FindHandler.Caller {

    private val repository: MangaListRepository = MangaListRepository(this)
    private val handler: FindHandler = FindHandler(this)

    private val mangaListAll = arrayListOf<Manga>()

    private fun loadData() {
        repository.callData()
    }

    override fun startLoading() {
        super.startLoading()
        loadData()
    }

    override fun stopLoading() {
        super.stopLoading()
        repository.cancelDataRequest()
    }

    override fun onGetData(data: ArrayList<Manga>) {
        updateMangaListAll(data)
        super.onGetData(data)
    }

    private fun updateMangaListAll(mangaList: ArrayList<Manga>) {
        mangaListAll.clear()
        mangaListAll.addAll(mangaList)
    }


    fun findData(pattern: String) {
        if (mangaListAll.isNotEmpty()) {
            if (pattern.isEmpty()) {
                postSearchResult(mangaListAll)
            } else {
                runSearchInNewThread(pattern)
                isLoading = true
            }
        }
    }

    override fun postSearchResult(mangaList: ArrayList<Manga>) {
        super.onGetData(mangaList)
    }

    private fun runSearchInNewThread(pattern: String) {

        val runnable = Runnable {
            handler.setResultList(mangaListAll.findForMatchesInTitle(pattern))
            handler.sendEmptyMessage(0)
        }

        val thread = Thread(runnable)
        thread.start()
    }

    override fun onCleared() {
        super.onCleared()
        handler.removeMessages(0)
    }
}
